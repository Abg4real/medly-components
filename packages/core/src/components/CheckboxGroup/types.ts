import { CheckboxSizes, FontSizes, FontWeights } from '@medly-components/theme';
import { HTMLProps, Omit, WithThemeProp } from '@medly-components/utils';
import { LabelPositions } from '../Label/types';

export interface Props extends Omit<HTMLProps<HTMLDivElement>, 'size' | 'onChange'>, WithThemeProp {
    /** Array of selected values */
    values?: any[];
    /** This method will be called with checked values */
    onChange: (values: any[]) => void;
    /** Options */
    options: Array<{ value: any; label: string; disabled?: boolean }>;
    /** Label */
    label?: string;
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** Disable all the checkboxes */
    disabled?: boolean;
    /** Label position */
    labelPosition?: LabelPositions;
    /** Label size */
    labelSize?: FontSizes;
    /** Label weight */
    labelWeight?: FontWeights;
    /** Label color */
    labelColor?: string;
    /** Name attribute of all the checkboxes */
    name?: string;
    /** Checkbox size */
    size?: CheckboxSizes;
    /** Show checkbox to select all options */
    showSelectAll?: boolean;
}
