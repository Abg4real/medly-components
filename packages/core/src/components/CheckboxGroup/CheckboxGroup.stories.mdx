import { CheckboxGroup } from './CheckboxGroup';
import { defaultTheme } from '@medly-components/theme';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './CheckboxGroup.stories';

<Meta title="Core/CheckboxGroup" component={CheckboxGroup} />

# CheckboxGroup

### Displays group of checkboxes

<Preview withToolbar>
    <Story name="Basic">{stories.Basic}</Story>
</Preview>

### Props

<Props of={CheckboxGroup} />

### CheckboxTheme

<Props of={stories.ThemeInterface} />

### CheckboxTheme Defaults

<code>{JSON.stringify(defaultTheme.checkbox, null, 2)}</code>
