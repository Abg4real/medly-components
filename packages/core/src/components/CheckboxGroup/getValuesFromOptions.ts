import { Props } from './types';

const getValuesFromOptions = (options: Props['options']): string[] =>
    options.reduce((acc, curr) => [...acc, ...(Array.isArray(curr.value) ? getValuesFromOptions(curr.value) : [curr.value])], []);

export default getValuesFromOptions;
