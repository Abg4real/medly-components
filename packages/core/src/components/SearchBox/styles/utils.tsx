import { SvgIcon } from '@medly-components/icons';
import { css } from '@medly-components/utils';
import { Props } from '../types';

export const getSearchBoxSize = ({ theme, size }: Props) => theme.searchBox.sizes[size].height;

export const getBorderAndBoxShadow = (color: string, boxShadow: string) => {
    return css`
        border-color: ${color};
        box-shadow: ${boxShadow};
    `;
};

export const getIconWrapperStyle = ({ theme, size }: Props) => {
    return css`
        width: ${getSearchBoxSize};
        min-width: ${getSearchBoxSize};
        height: auto;
        display: inline-flex;
        align-items: center;
        justify-content: center;
        ${SvgIcon} {
            cursor: pointer;
            font-size: ${theme.searchBox.sizes[size].iconSize};
        }
    `;
};
