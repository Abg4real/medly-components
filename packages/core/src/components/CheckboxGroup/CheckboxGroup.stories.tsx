import { boolean, select, text } from '@storybook/addon-knobs';
import React, { useState } from 'react';
import { CheckboxGroup } from './CheckboxGroup';
import { Props } from './types';

const labelPosition: Array<Props['labelPosition']> = ['left', 'right', 'top', 'bottom'];
const size: Array<Props['size']> = ['XS', 'S', 'M', 'L', 'XL'];

export const Basic = () => {
    const [values, setValues] = useState(['grapes']);

    return (
        <CheckboxGroup
            values={values}
            onChange={setValues}
            options={[
                { value: 'apple', label: 'Apple' },
                { value: 'orange', label: 'Orange' },
                { value: 'grapes', label: 'Grapes' },
                {
                    value: [
                        { value: 'custardApple', label: 'Custard Apple' },
                        { value: 'papaya', label: 'Papaya' }
                    ],
                    label: 'Group'
                }
            ]}
            disabled={boolean('Disabled', false)}
            fullWidth={boolean('Full Width', false)}
            showSelectAll={boolean('Show Select All', false)}
            label={text('Label', 'Fruits')}
            size={select('Size', size, 'S')}
            labelPosition={select('Label Position', labelPosition, 'top')}
        />
    );
};
