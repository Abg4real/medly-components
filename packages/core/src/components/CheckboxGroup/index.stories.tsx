import { boolean, select, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import Checkbox from '../Checkbox';
import CheckboxGroup from './CheckboxGroup';
import { Props } from './types';

const labelPosition: Array<Props['labelPosition']> = ['start', 'end', 'top', 'bottom'];
const size: Array<Props['size']> = ['XS', 'S', 'M', 'L', 'XL'];

storiesOf('Core', module).add('CheckboxGroup', () => (
    <CheckboxGroup
        label={text('Label', 'Fruits')}
        name={text('Name', 'fruits')}
        required={boolean('Required', false)}
        size={select('Size', size, 'S')}
        labelPosition={select('Label Position', labelPosition, 'top')}
    >
        <Checkbox label="Apple" />
        <Checkbox label="Orange" />
    </CheckboxGroup>
));
