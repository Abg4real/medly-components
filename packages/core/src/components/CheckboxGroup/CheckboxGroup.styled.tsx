import { styled } from '@medly-components/utils';
import FieldWithLabel from '../FieldWithLabel';
import { HelperText } from '../TextField/Styled';

export const CheckboxGroupWrapper = styled(FieldWithLabel)<{ error?: boolean; showSelectAll?: boolean; isIndented?: boolean }>`
    margin-left: ${({ isIndented }) => (isIndented ? '4rem' : '0')};
    ${FieldWithLabel.Field} {
        margin-left: ${({ showSelectAll }) => (showSelectAll ? '4rem' : '0')};
    }
    ${HelperText} {
        grid-area: description;
        color: ${({ error, theme }) => (error ? theme.colors.red[500] : theme.colors.grey[900])};
        margin: ${({ error, showSelectAll }) => (error && showSelectAll ? '0 0 1.2rem 4rem' : '0 0 1.2rem 0')};
    }
`;
