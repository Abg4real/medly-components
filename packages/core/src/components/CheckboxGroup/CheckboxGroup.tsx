import { useCombinedRefs, WithStyle } from '@medly-components/utils';
import React, { SFC, useCallback, useMemo } from 'react';
import Checkbox from '../Checkbox';
import { SelectorGroup } from '../Selectors';
import getValuesFromOptions from './getValuesFromOptions';
import { Props } from './types';

export const CheckboxGroup: SFC<Props> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const {
            id,
            values,
            onChange,
            options,
            size,
            label,
            labelVariant,
            labelWeight,
            showSelectAll,
            disabled,
            errorText,
            helperText,
            columns,
            required,
            parentHasError,
            ...wrapperProps
        } = props;

        const checkboxGroupId = useMemo(() => id || label, [id, label]),
            checkboxGroupRef = useCombinedRefs<HTMLDivElement>(ref, React.useRef(null)),
            hasError = useMemo(() => !!errorText || parentHasError, [errorText, parentHasError]),
            hasHelperOrErrorText = useMemo(() => !!(errorText || helperText), [errorText, helperText]),
            allChildValues = useMemo(() => getValuesFromOptions(options), [options]),
            areAllValuesSelected = useMemo(() => allChildValues.every(val => values.includes(val)), [options, values]),
            indeterminate = useMemo(() => !areAllValuesSelected && allChildValues.some(el => values.includes(el)), [
                areAllValuesSelected,
                allChildValues
            ]);

        const handleOptionClick = useCallback(
                (event: React.ChangeEvent<HTMLInputElement>) => {
                    event.stopPropagation();
                    const item = event.target.name,
                        isChecked = event.target.checked,
                        newValues = isChecked ? [...values, item] : values.filter(vl => vl !== item);
                    onChange(newValues);
                },
                [values, onChange]
            ),
            handleSelectAllClick = useCallback(() => {
                const newValues = areAllValuesSelected
                    ? values.filter(val => !allChildValues.includes(val))
                    : Array.from(new Set([...values, ...allChildValues]));
                onChange(newValues);
            }, [options, values, onChange]);

        return (
            <SelectorGroup.Wrapper
                ref={checkboxGroupRef}
                type="checkbox"
                id={`${checkboxGroupId}-wrapper`}
                aria-describedby={`${checkboxGroupId}-helper-text`}
                {...{ ...wrapperProps, hasHelperOrErrorText }}
            >
                {label &&
                    (showSelectAll ? (
                        <Checkbox
                            key="select-all"
                            label={label}
                            checked={areAllValuesSelected}
                            onChange={handleSelectAllClick}
                            indeterminate={indeterminate}
                            id={`${checkboxGroupId}-select-all`}
                            {...{ size, hasError, disabled, labelVariant, labelWeight }}
                        />
                    ) : (
                        <SelectorGroup.Label
                            id={`${checkboxGroupId}-label`}
                            type="checkbox"
                            required={required}
                            disabled={disabled}
                            textVariant={labelVariant}
                            textWeight={labelWeight}
                        >
                            {label}
                        </SelectorGroup.Label>
                    ))}
                {(errorText || helperText) && (
                    <SelectorGroup.HelperText
                        id={`${checkboxGroupId}-helper-text`}
                        type="checkbox"
                        disabled={disabled}
                        hasError={hasError}
                        isIndented={showSelectAll}
                    >
                        {errorText || helperText}
                    </SelectorGroup.HelperText>
                )}
                <SelectorGroup.Options id={`${checkboxGroupId}-options`} columns={columns} isIndented={showSelectAll}>
                    {options.map(option => {
                        return Array.isArray(option.value) ? (
                            <CheckboxGroup
                                key={option.label}
                                parentHasError={hasError}
                                id={`${option.label}-${checkboxGroupId}`}
                                {...{
                                    ...props,
                                    showSelectAll: option.showSelectAll ?? true,
                                    options: option.value,
                                    label: option.label,
                                    columns: option.columns,
                                    helperText: option.helperText,
                                    errorText: option.errorText
                                }}
                            />
                        ) : (
                            <Checkbox
                                size={size}
                                key={option.value}
                                name={option.value}
                                label={option.label}
                                checked={values.includes(option.value)}
                                onChange={handleOptionClick}
                                disabled={disabled || option.disabled}
                                hasError={hasError}
                                id={`${option.label}-${checkboxGroupId}`}
                            />
                        );
                    })}
                </SelectorGroup.Options>
            </SelectorGroup.Wrapper>
        );
    })
);

CheckboxGroup.displayName = 'CheckboxGroup';
CheckboxGroup.Style = SelectorGroup.Wrapper;
CheckboxGroup.defaultProps = {
    values: [],
    columns: 1,
    labelVariant: 'body1',
    labelWeight: 'Medium'
};
