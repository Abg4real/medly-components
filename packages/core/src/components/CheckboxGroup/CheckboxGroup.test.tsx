import { TestUtils } from '@medly-components/utils';
import React from 'react';
import CheckboxGroup from './CheckboxGroup';

describe('CheckboxGroup component', () => {
    afterEach(TestUtils.cleanup);

    it('should render correctly with all the props', () => {
        const { container } = TestUtils.render(
            <CheckboxGroup
                value={['apple']}
                onChange={jest.fn()}
                label="Fruits"
                options={[{ value: 'apple', label: 'Apple' }, { value: 'orange', label: 'Orange' }]}
            />
        );
        expect(container).toMatchSnapshot();
    });

    it('should render correctly when one of the option is selected', () => {
        const mockOnChange = jest.fn(),
            { getByText } = TestUtils.render(
                <CheckboxGroup
                    value={['apple']}
                    onChange={mockOnChange}
                    label="Fruits"
                    options={[{ value: 'apple', label: 'Apple' }, { value: 'orange', label: 'Orange' }]}
                />
            );
        TestUtils.fireEvent.click(getByText('Apple'));
        expect(mockOnChange).toHaveBeenCalledWith([]);
    });

    it('should select all when clicked on the select all option', () => {
        const mockOnChange = jest.fn(),
            { getByText } = TestUtils.render(
                <CheckboxGroup
                    showSelectAll
                    value={['apple']}
                    onChange={mockOnChange}
                    label="Fruits"
                    options={[{ value: 'apple', label: 'Apple' }, { value: 'orange', label: 'Orange' }]}
                />
            );
        TestUtils.fireEvent.click(getByText('Fruits'));
        expect(mockOnChange).toHaveBeenCalledWith(['apple', 'orange']);
    });

    it('should unselect all when all the options are already selected', () => {
        const mockOnChange = jest.fn(),
            { getByText } = TestUtils.render(
                <CheckboxGroup
                    showSelectAll
                    value={['apple', 'orange']}
                    onChange={mockOnChange}
                    label="Fruits"
                    options={[{ value: 'apple', label: 'Apple' }, { value: 'orange', label: 'Orange' }]}
                />
            );
        TestUtils.fireEvent.click(getByText('Fruits'));
        expect(mockOnChange).toHaveBeenCalledWith([]);
    });
});
