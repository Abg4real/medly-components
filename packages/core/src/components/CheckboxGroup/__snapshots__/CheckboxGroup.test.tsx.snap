// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup component should render correctly with all the default props 1`] = `
.c8 {
  font-size: 16px;
}

.c8 * {
  fill: #555;
}

.c0 {
  margin: 8px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c2 {
  margin: 0px 0px 0px 8px;
  color: #012040;
  cursor: pointer;
}

.c3 {
  width: 16px;
  height: 16px;
  position: relative;
  border: 1px solid #CCCCCC;
  background-color: #ffffff;
  border-radius: 15%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  width: 0;
  height: 0;
  z-index: 1;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 .c7 {
  width: 100%;
  height: 100%;
}

.c4 {
  opacity: 0;
  z-index: 2;
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
}

.c4:focus {
  outline: none;
}

.c4:checked ~ .c5 {
  width: calc(70%);
  height: calc(70%);
  -webkit-transition: width 0.2s ease,height 0.2s ease;
  transition: width 0.2s ease,height 0.2s ease;
}

<div>
  <div
    class="c0"
  >
    
    <div
      class="c1"
    >
      <label
        class="c2"
        for="Apple"
      >
        Apple
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="Apple"
          name=""
          type="checkbox"
        />
        <div
          class="c5 c6"
        >
          <svg
            class="c7 c8"
            height="1em"
            viewBox="0 0 18 14"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 11.17L1.83 7 .41 8.41 6 14 18 2 16.59.59z"
              fill="none"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
        for="Orange"
      >
        Orange
      </label>
      <div
        class="c3"
      >
        <input
          class="c4"
          id="Orange"
          name=""
          type="checkbox"
        />
        <div
          class="c5 c6"
        >
          <svg
            class="c7 c8"
            height="1em"
            viewBox="0 0 18 14"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 11.17L1.83 7 .41 8.41 6 14 18 2 16.59.59z"
              fill="none"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CheckboxGroup component should render correctly with all the props given 1`] = `
.c9 {
  font-size: 16px;
}

.c9 * {
  fill: #555;
}

.c2 {
  margin: 8px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c0 {
  margin: 8px 0px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  margin: 0px 0px 0px 8px;
  color: #012040;
  cursor: pointer;
}

.c1 {
  margin: 0px 0px 8px 0px;
  color: #012040;
}

.c1::after {
  color: red;
  content: ' *';
}

.c4 {
  width: 16px;
  height: 16px;
  position: relative;
  border: 1px solid #CCCCCC;
  background-color: #ffffff;
  border-radius: 15%;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  width: 0;
  height: 0;
  z-index: 1;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 .c8 {
  width: 100%;
  height: 100%;
}

.c5 {
  opacity: 0;
  z-index: 2;
  position: absolute;
  width: 100%;
  height: 100%;
  margin: 0;
  cursor: pointer;
}

.c5:focus {
  outline: none;
}

.c5:checked ~ .c6 {
  width: calc(70%);
  height: calc(70%);
  -webkit-transition: width 0.2s ease,height 0.2s ease;
  transition: width 0.2s ease,height 0.2s ease;
}

<div>
  <div
    class="c0"
  >
    <label
      class="c1"
      required=""
    >
      Fruits
    </label>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Apple"
      >
        Apple
      </label>
      <div
        class="c4"
      >
        <input
          class="c5"
          id="Apple"
          name="fruits"
          required=""
          type="checkbox"
        />
        <div
          class="c6 c7"
        >
          <svg
            class="c8 c9"
            height="1em"
            viewBox="0 0 18 14"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 11.17L1.83 7 .41 8.41 6 14 18 2 16.59.59z"
              fill="none"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="c2"
    >
      <label
        class="c3"
        for="Orange"
      >
        Orange
      </label>
      <div
        class="c4"
      >
        <input
          class="c5"
          id="Orange"
          name="fruits"
          required=""
          type="checkbox"
        />
        <div
          class="c6 c7"
        >
          <svg
            class="c8 c9"
            height="1em"
            viewBox="0 0 18 14"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 11.17L1.83 7 .41 8.41 6 14 18 2 16.59.59z"
              fill="none"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
